### DISCUSSION TITLE
$magewire dictionary variable not available for child blocks

### DISCUSSION THREAD

**Discussion by wpoortman** [magewirephp/magewire#53] [Ideas]
By default inside a component you always have the `$magewire` variable globally available as long as you are inside the phtml of the component itself. Now you get in trouble as soon as you render a child block who isn't a Magewire component on itself. The `$magewire` variable should still be available since theoretically you are still within the component's HTML even if it's in a child block.

**Comment by wpoortman** [comment]
So I've been working on this feature and I run into a situation where we need to make a decision whether the $magewire variable only is available for direct children or does it need to be available to all children until it reaches a following Magewire block.

I have an opinion, but I'll keep that to myself for now until a winning answer would be required.

Please let me know!

**Reply by speedupmate** [reply]
just have a setMagewire($magewire) against a childBlock , pretty common thing to do in m2?

**Reply by wpoortman** [reply]
He @speedupmate, thanks for the reply. My guess is that you need to work a bit more with Magewire to understand the feature. I wouldn't have asked or started the feature if a setMagewire would solve it 😉.

**Reply by speedupmate** [reply]
I'd say go to all child levels by default and have an option to break/force if a following Magewire  is injected in deeper level. Log a notice maybe in logs when this happens?

**Comment by ProxiBlue** [comment]
Not been using magewire as much as I would have liked, but, I will pipe up my opinion, as i think you have answered the question yourself already:

> since theoretically you are still within the component's HTML even if it's in a child block.

for myself, even if in nth child, I would consider that I am still within the components HTML, as the children are part of the whole that make up the  structure/display of the page.

**Reply by wpoortman** [reply]
Thanks for your reply Lucas!

Agree, but it can also create issues when (for some reason) you move that block out of the component. Being limited to the first child make it a bit more manageable.

Been doing some testing, works real well for both first-child-only and nested children. Still curious what other think when they get back from the weekend 😄.

**Comment by ProxiBlue** [comment]
> Agree, but it can also create issues when (for some reason) you move that block out of the component.

Should that be something you need to worry about, as the creator? If I developer move the block, it is my responsibility to ensure it is done correctly, and such a 'dependency' in the moved block is properly resolved.

IMO, by limiting the feature to 1st child only, you are limiting the overall richness/availability of the functionality.

**Reply by Vinai** [reply]
I find it really interesting that something like this would have been trivial in M1 to do, because "everything was a block" and "every block had exactly one parent", but in M2 it is a challenge because now we have containers and blocks and a block structure and 100x the code. Not a very helpful comment, but an interesting observation how generic data structures can be enabling.

**Comment by wpoortman** [comment]
Another brain fart is that you as a developer have the option to transform whatever block you want into a Magewire component. What I'm trying to say is that a child somewhere in the middle can consciously be changed where $magewire will be a different component from that point on. So when children heavily rely on the parent giving them the object and it suddenly changes, stuff can break.

Maybe a little far fetched and here we can also apply the rule that if a developer want's to, he should be aware of the consequences, but Magewire was also build to give developers the richness to transform whatever block into a Magewire component.

**Comment by wpoortman** [comment]
This will be fixed in V3 and will not be added to V1.
